src/link/remove-typename/removeTypenameFromVariables.ts

```diff
@@ -18,55 +18,81 @@ export interface RemoveTypenameFromVariablesOptions {
 export function removeTypenameFromVariables(
   options: RemoveTypenameFromVariablesOptions = Object.create(null)
 ) {
-  const { except } = options;
-
-  const trie = new Trie<typeof stripTypename.BREAK>(
-    canUseWeakMap,
-    () => stripTypename.BREAK
-  );
-
-  if (except) {
-    // Use `lookupArray` to store the path in the `trie` ahead of time. We use
-    // `peekArray` when actually checking if a path is configured in the trie
-    // to avoid creating additional lookup paths in the trie.
-    collectPaths(except, (path) => trie.lookupArray(path));
-  }
-
   return new ApolloLink((operation, forward) => {
+    const { except } = options;
     const { query, variables } = operation;
 
     if (!variables) {
       return forward(operation);
     }
 
-    if (!except) {
-      return forward({ ...operation, variables: stripTypename(variables) });
-    }
-
-    const variableDefinitions = getVariableDefinitions(query);
-
     return forward({
       ...operation,
-      variables: stripTypename(variables, {
-        keep: (variablePath) => {
-          const typename = variableDefinitions[variablePath[0]];
-
-          // The path configurations do not include array indexes, so we
-          // omit them when checking the `trie` for a configured path
-          const withoutArrayIndexes = variablePath.filter(
-            (segment) => typeof segment === 'string'
-          );
-
-          // Our path configurations use the typename as the root so we need to
-          // replace the first segment in the variable path with the typename
-          // instead of the top-level variable name.
-          return trie.peekArray([typename, ...withoutArrayIndexes.slice(1)]);
-        },
-      }),
+      variables: except
+        ? maybeStripTypenameUsingConfig(query, variables, except)
+        : stripTypename(variables),
     });
   });
 }
 
+function maybeStripTypenameUsingConfig(
+  query: DocumentNode,
+  variables: OperationVariables,
+  config: KeepTypenameConfig
+) {
+  const variableDefinitions = getVariableDefinitions(query);
+
+  return Object.fromEntries(
+    Object.entries(variables).map(([key, value]) => {
+      const typename = variableDefinitions[key];
+      const typenameConfig = config[typename];
+
+      return [
+        key,
+        typenameConfig
+          ? maybeStripTypename(value, typenameConfig)
+          : stripTypename(value),
+      ];
```

Comment:
This part of the code might not be critical for performance now, but one fun thing about the `Object.fromEntries(Object.entries().map(...))` pattern is you can reuse the `[key, value]` entry arrays without allocating a new array:
```suggestion
    Object.entries(variables).map(keyVal => {
      const [key, value] = keyVal;
      const typename = variableDefinitions[key];
      const typenameConfig = config[typename];

      keyVal[1] = typenameConfig
          ? maybeStripTypename(value, typenameConfig)
          : stripTypename(value);
      
      return keyVal;
    });
```

Reply:
Good call! Will update.